warehouse/utils/project.py

```diff
@@ -92,6 +98,64 @@ def prohibit_and_remove_project(
         remove_project(project, request, flash=flash)
 
 
+def quarantine_project(project: Project, request, flash=True) -> None:
+    """
+    Quarantine a project. Reversible action.
+    """
+    project.lifecycle_status = LifecycleStatus.QuarantineEnter
+    project.lifecycle_status_note = f"Quarantined by {request.user.username}."
+
+    project.record_event(
+        tag=EventTag.Project.ProjectQuarantineEnter,
+        request=request,
+        additional={"submitted_by": request.user.username},
+    )
+
+    request.db.add(
+        JournalEntry(
```

Comment:
Thinking about this, I'm not sure we want to emit these journal entries until we are making other user facing changes.

I'm unsure how much we want to broadcast at all. I'm thinking about the scenario where a false positive is quarantined and then released, and the repercussions that could have.

For now... I think its fine but it raises some questions for me.

Reply:
I can definitely remove these until we have user-facing behaviors (which should be the next PR in this effort). I don't imagine we'll be using the feature in production until it actively performs the actions described.

>  I'm thinking about the scenario where a false positive is quarantined and then released, and the repercussions that could have.

Definitely a concern. From poking at `bandersnatch`, it appears that this will have the intended impact - the code tolerates a 404 after having synced the project once. If the user has "cleanup" mode, it'll be removed from the mirror, which is also probably fine - since that has the intended affect as well.
And if the project appears after exiting quarantine, the entry updating the changelog will cause the mirror to check in again. If it wasn't cleaned up, no harm, no foul. If it was cleaned up, it'll re-sync.

@dstufft had thoughts on omission leading to mirror inconsistencies - maybe you can detail those here?

Reply:
Discussion on the implications of publishing "quarantined" and "un quarantined" events led me to think its better t one explicit.

Sounds like local testing shows that bandersnatch (at least latest) handles these events well, so I think I'm resolved here.

Reply:
If we're hiding the package from the `/simple/` index, then that _is_ a user facing change yea?

The issue becomes a sequence of events like this:

1. User uploads a package, triggering a journal entry.
2. File gets hidden from `/simple/` without a journal entry.
3. A mirror asks PyPI for changes since some serial, which includes (1), and gets the package.
4. Mirror syncs the package, missing the hidden file (because it's hidden), and marks itself as having synced past (1).
5. File gets unhidden, but no journal entry so mirrors don't sync the Package.

Or alternatively:

1. User uploads a package, triggering a journal entry.
2. Mirror syncs past (1).
3. File gets hidden on PyPI without a journal entry.
4. Mirror continues to serve file, causing different behavior between PyPI and the mirror.
   - This could be combined with the above, causing some mirrors to have it and some not, but all of them think they're at the latest version.


One of the implicit promises we've made is that any meaningful changes to the `/simple/` pages will come with an accompanying entry in the journals table, as that's how the mirrors know that the contents of those pages have changed and they need to resync them.